src/cpu/testers/spatter_gen/spatter_gen.cc

```diff
@@ -0,0 +1,584 @@
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "cpu/testers/spatter_gen/spatter_gen.hh"
+
+#include "base/cprintf.hh"
+#include "base/random.hh"
+#include "cpu/testers/spatter_gen/enums.hh"
+#include "debug/SpatterGen.hh"
+#include "debug/SpatterKernel.hh"
+#include "sim/sim_exit.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+
+using enums::SpatterKernelTypeStrings;
+
+SpatterGen::SpatterGen(const Params& params):
+    ClockedObject(params),
+    port(this, name() + ".port"),
+    requestorId(params.system->getRequestorId(this)),
+    numPendingMemRequests(0),
+    stats(this),
+    intRegFileSize(params.int_regfile_size), intRegUsed(0),
+    fpRegFileSize(params.fp_regfile_size), fpRegUsed(0),
+    requestGenLatency(params.request_gen_latency),
+    requestGenBandwidth(params.request_gen_bandwidth),
+    firstGeneratorAvailableTime(0),
+    nextGenEvent([this](){ processNextGenEvent(); }, name() + ".GenEvent"),
+    requestBufferSize(params.request_buffer_size),
+    sendBandwidth(params.send_bandwidth),
+    firstPortAvailableTime(0),
+    nextSendEvent([this](){ processNextSendEvent(); }, name() + ".SendEvent")
+{
+    generatorBusyUntil.resize(requestGenBandwidth);
+    for (int i = 0; i < requestGenBandwidth; i++) {
+        generatorBusyUntil[i] = 0;
+    }
+    portBusyUntil.resize(sendBandwidth);
+    for (int i = 0; i < sendBandwidth; i++) {
+        portBusyUntil[i] = 0;
+    }
+}
+
+Port&
+SpatterGen::getPort(const std::string& if_name, PortID idx)
+{
+    if (if_name == "port") {
+        return port;
+    } else {
+        return ClockedObject::getPort(if_name, idx);
+    }
+}
+
+void
+SpatterGen::addKernel(
+    uint32_t id, uint32_t delta, uint32_t count,
+    SpatterKernelType type,
+    size_t index_size, Addr base_index_addr,
+    size_t value_size, Addr base_value_addr,
+    std::vector<uint32_t> indices
+)
+{
+    DPRINTF(
+        SpatterGen,
+        "%s: Adding kernel with id: %d, delta: %d, count: %d, type: %s.\n",
+        __func__, id, delta, count, SpatterKernelTypeStrings[type]
+    );
+    SpatterKernel new_kernel(id, delta, count, type,
+                            index_size, base_index_addr,
+                            value_size, base_value_addr
+                            );
+    new_kernel.setIndices(indices);
+    kernels.push(new_kernel);
+}
+
+void
+SpatterGen::startup()
+{
+    scheduleNextGenEvent(curTick());
+}
+
+void
+SpatterGen::checkForSimExit()
+{
+    bool no_pending_mem = numPendingMemRequests == 0;
+    bool no_queued_req = requestBuffer.empty();
+    bool no_index = receiveBuffer.empty();
+    bool no_kernel = kernels.empty();
+    if (no_pending_mem && no_queued_req && no_index && no_kernel)
+    {
+        exitSimLoop(
+            csprintf("SpatterGen %s done generating requests.", name()),
+            0,
+            nextCycle()
+        );
+    }
```

Comment:
Is this how the other generators work? I thought if they had nothing to do, they would just not schedule any events and you would get a `MaxTick` exit. I would like all generators to have the same behavior.

Reply:
What do you think about this?

Reply:
You are right, this behavior is different. However, I remember having much trouble with using bandwidth stats with that behavior.